File: CB_TimberFraming/configure.rb
# load "G:/My Drive/TF/Sketchup/Rubies/CB_TimberFraming/CB_TimberFraming/configure.rb"


module CB_TF
  def CB_TF.configure
    dialog = UI::HtmlDialog.new(
      {
        :dialog_title => "Timber Framing Extensions Configuration",
        :scrollable => true,
        :resizable => false,
        :width => 700,
        :height => 750,
        :left => 200,
        :top => 200,
        :style => UI::HtmlDialog::STYLE_DIALOG
      })

    path = File.dirname(__FILE__)

    replacements = {  
      config_json: {
        company_name: Sketchup.read_default("TF", "company_name", ""),
        side_spacing: Sketchup.read_default("TF", "side_spacing", "30"),
        dir_labels: Sketchup.read_default("TF", "dir_labels", true),
        roundup: Sketchup.read_default("TF", "roundup", true),
        list_file_format: Sketchup.read_default("TF", "list_file_format", "C"),
        list_by_tag: Sketchup.read_default("TF", "list_by_tag", false),
        metric: Sketchup.read_default("TF", "metric", false),
        roll: Sketchup.read_default("TF", "roll", false),
        min_extra_timber_length: Sketchup.read_default("TF", "min_extra_timber_length", "24"),
        qty: Sketchup.read_default("TF", "qty", true),
        sq_x_pos: Sketchup.read_default("TF", "sq_x_pos", "16.5").to_f,
        sq_y_pos: Sketchup.read_default("TF", "sq_y_pos", "1.125").to_f,
        sq_font_size: Sketchup.read_default("TF", "sq_font_size", "18"),
        sq_rotate: Sketchup.read_default("TF", "sq_rotate", "270").to_i,
        sq_bold: Sketchup.read_default("TF", "sq_bold", true),
        vp2dx: Sketchup.read_default("TF", "vp2dx", 1.125),
        vp2dy: Sketchup.read_default("TF", "vp2dy", 0.375),
        vp2dw: Sketchup.read_default("TF", "vp2dw", 13.0),
        vp2dh: Sketchup.read_default("TF", "vp2dh", 8.0), 
        vp3dx: Sketchup.read_default("TF", "vp3dx", 14.125),
        vp3dy: Sketchup.read_default("TF", "vp3dy", 0.375),
        vp3dw: Sketchup.read_default("TF", "vp3dw", 2.0),
        vp3dh: Sketchup.read_default("TF", "vp3dh", 10.0),
        t3do: Sketchup.read_default("TF", "t3do", "V"),
      }.to_json,
      javascript: File.read(File.join(path, "configure.js")),
      stylesheet: File.read(File.join(path, "configure.css"))
    }

    html = File.read(File.join(path, "configure.html"))
    dialog.set_html(html % replacements)
    dialog.add_action_callback("tf_save_config") { |action_context, data|
      puts "saving config data: \n#{data}"
      dialog.close()

      Sketchup.write_default("TF", "company_name", data["company_name"])      
      data["dir_labels"] ? Sketchup.write_default("TF", "dir_labels", 1) : Sketchup.write_default("TF", "dir_labels", 0)
      data["roundup"] ? Sketchup.write_default("TF", "roundup", 1) : Sketchup.write_default("TF", "roundup", 0)
      Sketchup.write_default("TF", "list_file_format", data["list_file_format"])
      data["list_by_tag"] ? Sketchup.write_default("TF", "list_by_tag", 1) : Sketchup.write_default("TF", "list_by_tag", 0)
      data["metric"] ? Sketchup.write_default("TF", "metric", 1) : Sketchup.write_default("TF", "metric", 0)
      data["roll"] ? Sketchup.write_default("TF", "roll", 1) : Sketchup.write_default("TF", "roll", 0)
      Sketchup.write_default("TF", "side_spacing", data["side_spacing"].to_i )
      Sketchup.write_default("TF", "min_extra_timber_length", data["min_extra_timber_length"].to_i )
      data["qty"] ? Sketchup.write_default("TF", "qty", 1) : Sketchup.write_default("TF", "qty", 0)
      Sketchup.write_default("TF", "sq_x_pos", data["sq_x_pos"])
      Sketchup.write_default("TF", "sq_y_pos", data["sq_y_pos"])
      Sketchup.write_default("TF", "sq_font_size", data["sq_font_size"])
      Sketchup.write_default("TF", "sq_rotate", data["sq_rotate"])
      Sketchup.write_default("TF", "sq_bold", data["sq_bold"])
      Sketchup.write_default("TF", "vp2dx", data["vp2dx"])
      Sketchup.write_default("TF", "vp2dy", data["vp2dy"])
      Sketchup.write_default("TF", "vp2dw", data["vp2dw"])
      Sketchup.write_default("TF", "vp2dh", data["vp2dh"])
      Sketchup.write_default("TF", "vp3dx", data["vp3dx"])
      Sketchup.write_default("TF", "vp3dy", data["vp3dy"])
      Sketchup.write_default("TF", "vp3dw", data["vp3dw"])
      Sketchup.write_default("TF", "vp3dh", data["vp3dh"])
      Sketchup.write_default("TF", "t3do", data["t3do"])
    }
    dialog.add_action_callback("tf_cancel") { |action_context|
      puts "User canceled"
      dialog.close()
    }
    dialog.show_modal

  end
end

File: CB_TimberFraming/configure.js
document.onreadystatechange = function () {
  if (document.readyState === 'complete') {
    var ci = document.getElementById('company_name');
    ci.value = data.company_name;
    
    ci = document.getElementById('side_spacing');
    ci.value = data.side_spacing;

    if (data.dir_labels == true){
      ci = document.getElementById('dir_labelsY');
      ci.checked = true;
    } else {
      ci = document.getElementById('dir_labelsN');
      ci.checked = true;
    }

    if (data.roundup == true){
      ci = document.getElementById('roundupY');
      ci.checked = true;
    } else {
      ci = document.getElementById('roundupN');
      ci.checked = true;
    }

    if (data.list_by_tag == true){
      ci = document.getElementById('list_by_tagY');
      ci.checked = true;
    } else {
      ci = document.getElementById('list_by_tagN');
      ci.checked = true;
    }
    
    if (data.metric == true){
      ci = document.getElementById('metric');
      ci.checked = true;
    } else {
      ci = document.getElementById('english');
      ci.checked = true;
    }
    
    if (data.list_file_format == "C"){
      ci = document.getElementById('lff_csv');
      ci.checked = true;
    } else if (data.list_file_format == "T") {
      ci = document.getElementById('lff_txt');
      ci.checked = true;
    } else {
      ci = document.getElementById('lff_xls');
      ci.checked = true;
    }

    if (data.roll == true){
      ci = document.getElementById('roll');
      ci.checked = true;
    } else {
      ci = document.getElementById('unwrap');
      ci.checked = true;
    }      

    ci = document.getElementById('min_extra_timber_length');
    ci.value = data.min_extra_timber_length;

    if (data.qty == true){
      ci = document.getElementById('qtyY');
      ci.checked = true;
      show_qty_config();
    } else {
      ci = document.getElementById('qtyN');
      ci.checked = true;
      hide_qty_config();
    }

    ci = document.getElementById('sq_x');
    ci.value = data.sq_x_pos;

    ci = document.getElementById('sq_y');
    ci.value = data.sq_y_pos;

    ci = document.getElementById('sq_font_size');
    ci.value = data.sq_font_size;

    ci = document.getElementById('sq_rotate');
    ci.value = data.sq_rotate;

    if (data.sq_bold == true){
      ci = document.getElementById('sq_boldY');
      ci.checked = true;
    } else {
      ci = document.getElementById('sq_boldN');
      ci.checked = true;
    }    
    
    ci = document.getElementById('vp2dx');
    ci.value = data.vp2dx;
    
    ci = document.getElementById('vp2dy');
    ci.value = data.vp2dy;
    
    ci = document.getElementById('vp2dw');
    ci.value = data.vp2dw;
    
    ci = document.getElementById('vp2dh');
    ci.value = data.vp2dh;
        
    ci = document.getElementById('vp3dx');
    ci.value = data.vp3dx;

    ci = document.getElementById('vp3dy');
    ci.value = data.vp3dy;

    ci = document.getElementById('vp3dw');
    ci.value = data.vp3dw;

    ci = document.getElementById('vp3dh');
    ci.value = data.vp3dh;

    if (data.t3do == "H"){
      ci = document.getElementById('t3d_horizontal');
      ci.checked = true;
    } else if (data.t3do == "V") {
      ci = document.getElementById('t3d_vertical');
      ci.checked = true;
    } else {
      ci = document.getElementById('t3d_none');
      ci.checked = true;
    }
  }
}

function show_qty_config(){
  console.log("show");
  document.getElementById('size_qty_div').style.visibility = 'visible';
}

function hide_qty_config(){
  console.log("hide");
  document.getElementById('size_qty_div').style.visibility = 'hidden';
}

function save_data() {
  var ci = document.getElementById('company_name');
  data.company_name = ci.value;

  ci = document.getElementById('side_spacing');
  data.side_spacing = ci.value;

  ci = document.getElementById('dir_labelsY');
  data.dir_labels = ci.checked;

  ci = document.getElementById('roundupY');
  data.roundup = ci.checked;

  ci = document.getElementById('list_by_tagY');
  data.list_by_tag = ci.checked;            
  
  ci = document.getElementById('metric');
  data.metric = ci.checked;            
  
  ci = document.getElementById('lff_csv');
  if (ci.checked){
    data.list_file_format = "C";
  };
  ci = document.getElementById('lff_txt');
  if (ci.checked){
    data.list_file_format = "T";
  };
  ci = document.getElementById('lff_xls');
  if (ci.checked){
    data.list_file_format = "X";
  };

  ci = document.getElementById('roll');
  data.roll = ci.checked;            

  ci = document.getElementById('min_extra_timber_length');
  data.min_extra_timber_length = ci.value;

  ci = document.getElementById('qtyY');
  data.qty = ci.checked;

  ci = document.getElementById('sq_x');
  data.sq_x_pos = ci.value;

  ci = document.getElementById('sq_y');
  data.sq_y_pos = ci.value;

  ci = document.getElementById('sq_font_size');
  data.sq_font_size = ci.value;

  ci = document.getElementById('sq_rotate');
  data.sq_rotate = ci.value;

  ci = document.getElementById('sq_boldY');
  data.sq_bold = ci.checked;  

  ci = document.getElementById('vp2dx');
  data.vp2dx = ci.value;

  ci = document.getElementById('vp2dy');
  data.vp2dy = ci.value;

  ci = document.getElementById('vp2dw');
  data.vp2dw = ci.value;

  ci = document.getElementById('vp2dh');
  data.vp2dh = ci.value;

  ci = document.getElementById('vp3dx');
  data.vp3dx = ci.value;

  ci = document.getElementById('vp3dy');
  data.vp3dy = ci.value;

  ci = document.getElementById('vp3dw');
  data.vp3dw = ci.value;

  ci = document.getElementById('vp3dh');
  data.vp3dh = ci.value;

  ci = document.getElementById('t3d_vertical');
  if (ci.checked){
    data.t3do = "V";
  };
  ci = document.getElementById('t3d_horizontal');
  if (ci.checked){
    data.t3do = "H";
  };
  ci = document.getElementById('t3d_none');
  if (ci.checked){
    data.t3do = "N";
  };

  sketchup.tf_save_config(data);
}
